### ISSUE TITLE
Can't have rpcconnect on bitcoin.conf

### ISSUE THREAD

**Issue by bu-tester** [BitcoinUnlimited/BitcoinUnlimited#990]
I've my bitcoin rpc server listening on an alternate ip and when starting bitcoind, it'll stop with an error about having rpcconnect in the bitcoin.conf file.

So, currently I've only one of two options:
a) invoke bitcoin-cli w/ -rpcconnect=x.x.x.x, on every _single_ call, or
b) start bitcoind w/o rpcconnect on bitcoin.conf, wait, and only after, edit it to add back the rpcconnect setting, with which now I can invoke bitcoin-cli again, without having to manually specify the rpc ip through -rpcconnect=x.x.x.x

I'm assuming, that there's a whitelist of valid commands within bitcoin.conf, that bitcoind sanitizes on startup, however, it's forgetting that bitcoin-cli can **require** rpcconnect to connect into an alternate ip (and they both share the very same conf file).

Please advise if additional info is required. I've noticed this about +/- six months ago and have (wrongly) assumed to be an easy to spot issue - alas, it has not (yet) been fixed/noticed, for the last couple of releases.

Thanks

**Comment by sickpig** [comment]
> I've my bitcoin rpc server listening on an alternate ip 

how do you achieve this set up? via `rpcbind` parameter? 
 
> and when starting bitcoind, it'll stop with an error about having rpcconnect in the bitcoin.conf file.

what is the actual error message?

**Comment by bu-tester** [comment]
$ bitcoind -daemon
Error reading configuration file: unrecognized option 'rpcconnect'

yes, I have "rpcbind=127.0.0.253" on bitcoin.conf (I share the same port across different implementations, just binding through distinct ip's)

**Comment by sickpig** [comment]
> Error reading configuration file: unrecognized option 'rpcconnect'

this is strange cause `rpcconnect` is a legit conf parameter and such kind of error is thrown by BU only for non defined parameters. Would you mind share your entire `bitcoin.conf` file? 

> yes, I have "rpcbind=127.0.0.253" on bitcoin.conf (I share the same port across different implementations, just binding through distinct ip's)

two more questions:

- so you run multiple implementations on the same machine. do you run them under different users? 

- which version of BU are you testing? 1.2.0.1 or something else?

**Comment by bu-tester** [comment]
Yes, it *is* strange.

There's nothing fancy on the config, besides a custom RPC bind IP
----------------------------
$ cat bitcoin.conf
bind=127.0.0.253:9333
connect=x.x.x.x:yyyy

rpcbind=127.0.0.253
rpcport=9332
#rpcconnect=127.0.0.253
rpcallowip=127.0.0.0/24

rpcuser=some_user
rpcpassword=some_pass

txindex=1
disablewallet=1

listen=1
----------------------------

Not that it should be relevant, but yes, I have run multiple implementations concurrently. Initially XT, then Unlimited, Classic, ABC, Core. Currently however, I'm only using Unlimited to track Cash and Legacy chains (each on it's own isolated user). Latest BU (1.2.0.1) - and 1.0.3.0 (for the legacy one).

**Comment by ptschip** [comment]
i don't beleive rpcconnect is a config option , not for the node, the
rpcconect is a bitcoincli option...the other rpc options are there int
the config to allow rpcconnect when called from bitcoincli to access
your node.

**Comment by bu-tester** [comment]
True, it's not a config option for the node (bitcoind). It is for the bitcoincli, as you mention. However, **both** do share the same configuration file (bitcoin.conf). So, why is bitcoind stomping on bitcoincli requirements? Will they now need to use two different default configuration files?

**Comment by ptschip** [comment]
You're right and after examining the code i don't see why it's not being
added , the code clearly shows that it was intended to be added in.

@awemany  can you take a quick look at this, if i recall this was a
feature you coded up...it seems the rpcconnect option is not getting
included in m_args but i can see at the end of allowed_args.cpp that
those options are the first to get included...it's not entirely obvious
to me why this is?

**Comment by awemany** [comment]
@ptschip: No, that wasn't me.

But from looking at code, the *only* place that rpcconnect seems to be allowed  for is bitcoin-cli, though?
So isn't the behavior of bitcoind complaining expected? I'd still consider that a bug, of course.

I think one simply needs to add a dummy argument 'rpcconnect' to bitcoind in allowed_args.cpp to make it ignore that argument?

**Comment by ptschip** [comment]
i saw that but right at the bottom all the args get merged..so all args
should be possible to enter into bitcoin.conf but clearly that isn't the
case.

**Comment by ptschip** [comment]
@bu-tester just opened PR 998 which fixes this issue..thanks for reporting it.

**Comment by sickpig** [comment]
#998 merged by @gandrewstone, @bu-tester does it fix it for you?

**Comment by bu-tester** [comment]
I'll try again, but it now seems to complain about "disablewallet"?

$bitcoind
Error reading configuration file: unrecognized option 'disablewallet'

$ bitcoind --version
Bitcoin Unlimited Cash Edition Daemon version v1.2.0.1-924c27a8b

did I got the commit too early?

**Comment by ptschip** [comment]
disablewallet works for me... did you do "disablewallet=1"  you have to
add the value of the key, you can't just do "disablewallet" or
"disablewallet="

**Comment by bu-tester** [comment]
yep, I've "disablewallet=1" in bitcoin.conf
I'm doing a full rebuild now (make clean), had made just a git pull and make before.
(it's a slow process on that test/toy box)

**Comment by bu-tester** [comment]
Same error,

$ bitcoind
Error reading configuration file: unrecognized option 'disablewallet'

content of bitcoin.conf
$ cat bitcoin.conf
#bind=...
bind=127.0.0.253:9333
connect=x:y

rpcbind=127.0.0.253
rpcport=9332
#rpcconnect=...
rpcallowip=127.0.0.0/24

rpcuser=some_user
rpcpassword=some_pass

txindex=1
disablewallet=1

#proxy=...
listen=1
#onlynet=onion
#listenonion=0
#discover=0
##torcontrol=....
##torpassword=....
#externalip=...


#addnode=...
#addnode=...

# oom-killer countermeasures
##maxmempool=20
##use-thinblocks=0

dbcache=450
maxmempool=450


Ideas?

**Comment by bu-tester** [comment]
Heck! Even with just a single line on bitcoin.conf of "disablewallet=1" it trows the error.

$ git status
On branch dev
Your branch is up-to-date with 'origin/dev'.
nothing to commit, working tree clean

**Comment by ptschip** [comment]
thats odd...i'm on the latest dev, running on windows10.  are you
running QT or bitcoind version?

**Comment by ptschip** [comment]
try putting something in for rpcconnect ... that should pass and at
least validate you're running the current dev.

**Comment by bu-tester** [comment]
Adding just the rcpconnect=x besides disablewallet=1 bares no change, same error.
This is on armv71/armhf, but shouldn't matter. What's maybe relevant it's that the configure had been most likely done with --disable-wallet IIRC, so the build generated by make won't have wallet support builtin. If that's the culprit it's a new issue not seen before.
Nevertheless, commenting out disablewallet=1 is enough to get bitcoind working properly along rpcconnect. Now I can finally just start bitcoind and call bitcoin-cli again.

**Comment by bu-tester** [comment]
I can confirm that bitcoind does **not** complain about "disablewallet=1" on bitcoin.conf, when wallet support is built in, as in:

$ ./configure --without-miniupnpc --enable-hardening --disable-tests --disable-bench --with-incompatible-bdb --disable-asm
$ make -j3

it starts without errors while parsing the bitcoin.conf file (tested it while using an alternate conf with a single line of: "disablewallet=1"

$ cat bitcoin.conf2
disablewallet=1

$ bitcoind -daemon -conf=/full_path_to/bitcoin.conf2
Bitcoin server starting

However, it does fail if built without wallet:
$ ./configure --disable-wallet --without-miniupnpc --enable-hardening --disable-tests --disable-bench
$make -j3

$ bitcoind -daemon -conf=/full_path_to/bitcoin.conf2
Error reading configuration file: unrecognized option 'disablewallet'

Sorry for going on a tangent here (or two), but building on debian:stretch armhf, currently I can't compile wallet support (w/ libdb++-dev, not the old libdb4.8 - didn't test that) **if** I don't also disable asm -hence the --disable-asm flag (build error, will probably report later).

And, if I may ask (on a way unrelated question), why can't I pass relative paths to  bitcoind conf, as in bitcoind -conf=~/.bitcon/bitcoin2.conf (which get's completely ignored - or seems to, just no warning at all, probably defaults to plain bitcoin.conf if there's one? no, that's also not the case - because - it won't complain about disablewallet=1, if it's there), it will only accept the full path, as in bitcoind -conf=/full_path_to/bitcoin.conf2
This has probably been beaten to death elsewhere long ago, but I hadn't followed closely.

At least,, the rpcconnect specific issue seems fixed. The disablewallet=1 issue on a build without wallet support was not present on the release builds, only current dev (can try to bisect the breaking commit, but it's a slow box and there surely faster ways to catch it).

**Comment by bu-tester** [comment]
Disregard any prior mentioned issues with building with libdb++ and asm. It's now working fine - maybe some upstream update for the toolchain fixed it (or, I might had simply forgot to call ./autogen.sh, on which btw, there seems to be a typo on the installation notes of https://www.bitcoinunlimited.info/download, where it suggests to call ./autogen-sh - instead of ./autogen.sh).

Regarding ".configure --disable-wallet", it really won't allow later to have "disablewallet=1" within bitcoin.conf. Both bitcoind and bitcoin-qt won't start:

$ cat bitcoin.conf
disablewallet=1

$ bitcoind -daemon
Error reading configuration file: unrecognized option 'disablewallet'

$ bitcoin-qt
Error: Cannot parse configuration file: unrecognized option 'disablewallet'. Only use key=value syntax.

While it's true that builtin support for a wallet is not included in the binaries (per configure flag before make), it should at most be a benign warning, not a fatal error that blocks the daemon start.
